File: examples/go/token-nft/NFTExample.tsop.go
package contract

import "tsop"

type SimpleNFT struct {
	tsop.StatefulSmartContract
	Owner    tsop.PubKey     // stateful
	TokenId  tsop.ByteString `tsop:"readonly"` // immutable: unique token identifier
	Metadata tsop.ByteString `tsop:"readonly"` // immutable: token metadata URI/hash
}

func (c *SimpleNFT) Transfer(sig tsop.Sig, newOwner tsop.PubKey, outputSatoshis tsop.Bigint) {
	tsop.Assert(tsop.CheckSig(sig, c.Owner))
	// AddOutput(satoshis, owner) -- single mutable prop
	c.AddOutput(outputSatoshis, newOwner)
}

func (c *SimpleNFT) Burn(sig tsop.Sig) {
	// Only owner can burn
	tsop.Assert(tsop.CheckSig(sig, c.Owner))
	// No AddOutput and no state mutation = token destroyed
}

File: examples/rust/token-nft/NFTExample.tsop.rs
use tsop::prelude::*;

#[tsop::contract]
pub struct SimpleNFT {
    pub owner: PubKey, // stateful
    #[readonly]
    pub token_id: ByteString,
    #[readonly]
    pub metadata: ByteString,
}

#[tsop::methods(SimpleNFT)]
impl SimpleNFT {
    #[public]
    pub fn transfer(&mut self, sig: &Sig, new_owner: PubKey, output_satoshis: Bigint) {
        assert!(check_sig(sig, &self.owner));
        self.add_output(output_satoshis, new_owner);
    }

    #[public]
    pub fn burn(&self, sig: &Sig) {
        assert!(check_sig(sig, &self.owner));
        // No add_output = token destroyed
    }
}
